api/basic/add-pin.ts

```diff
@@ -0,0 +1,178 @@
+import { DynamoDBClient, ConditionalCheckFailedException } from '@aws-sdk/client-dynamodb'
+import { DynamoDBDocumentClient, GetCommand, PutCommand } from '@aws-sdk/lib-dynamodb'
+import { SQSClient, SendMessageCommand } from '@aws-sdk/client-sqs'
+import { APIGatewayProxyEventV2 } from 'aws-lambda'
+import { ClusterAddResponse, Pin, Response } from './schema.js'
+import { CID } from 'multiformats/cid'
+import { Multiaddr } from 'multiaddr'
+import { nanoid } from 'nanoid'
+
+interface UpsertPinInput {
+  cid: string
+  dynamo: DynamoDBClient
+  table: string
+}
+
+interface AddToQueueInput {
+  cid: string
+  origins: string[]
+  bucket: string
+  sqs: SQSClient
+  queueUrl: string
+}
+
+// type AddPinInput = AddToQueueInput & UpsertPinInput
+interface AddPinInput extends UpsertPinInput, AddToQueueInput {}
+
+/**
+ * AWS API Gateway handler for POST /pin/${cid}?&origins=${multiaddr},${multiaddr}
+ * Collect the params and delegate to addPin to do the work
+ *
+ * We provide responses in Payload format v2.0
+ * see: https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api-develop-integrations-lambda.html#http-api-develop-integrations-lambda.proxy-format
+ */
+export async function handler (event: APIGatewayProxyEventV2): Promise<Response> {
+  const {
+    TABLE_NAME: table = '',
+    BUCKET_NAME: bucket = '',
+    QUEUE_URL: queueUrl = '',
+    CLUSTER_BASIC_AUTH_TOKEN: token = '',
+    // set for testing
+    SQS_ENDPOINT: sqsEndpoint = undefined,
+    DYNAMO_DB_ENDPOINT: dbEndpoint = undefined
+  } = process.env
+
+  if (event.headers.authorization !== `Basic ${token}`) {
+    return { statusCode: 401, body: { error: { reason: 'UNAUTHORIZED' } } }
+  }
+
+  const sqs = new SQSClient({ endpoint: sqsEndpoint })
+  const dynamo = new DynamoDBClient({ endpoint: dbEndpoint })
+  const cid = event.pathParameters?.cid ?? ''
+  const origins = event.queryStringParameters?.origins?.split(',') ?? []
+  try {
+    const res = await addPin({ cid, origins, bucket, sqs, queueUrl, dynamo, table })
+    return { ...res, body: JSON.stringify(res.body) }
+  } catch (error) {
+    console.log(error)
```

Comment:
would be nice to adopt a logger with `warn`, `error`, `info`... Not needed in scope of this PR, but should exist for MVP

Reply:
Also note that Errors from `addPin` won't be logged here. I would change them to throw an error instance with the reason and then in this catch would check if is instance of INTERNAL_SERVER_ERROR

Reply:
What's our preferred logger. Is it Pino?

Reply:
I remember seeing Pino usage on some of Elastic IPFS repos, would like to keep consistency. Perhaps worth asking the rest of the folks working on top of AWS what they are using

Reply:
Will do as another PR https://github.com/olizilla/pickup/issues/22